demonstrations/tutorial_testing_symmetry.py

```diff
@@ -0,0 +1,465 @@
+Wilde (2022) <https://arxiv.org/pdf/2203.10017.pdf>`__ for testing the
+symmetries of a Hamiltonian. We’ll be able to determine whether a system
+has a finite group of symmetries :math:`G`, and if not, by how much the
+symmetry is violated.
+
+Background
+--------------
+
+We will encode symmetries into a `finite group
+<https://en.wikipedia.org/wiki/Finite_group>`__. This is an
+algebraic structure consisting of transformations :math:`g`, which act
+on the Hilbert space :math:`\mathcal{H}` of our system in the form of
+unitary operators :math:`U(g)` for :math:`g \in G`. More formally,
+given any two elements :math:`g_1, g_2\in G`, there is a *product* :math:`g_1 \circ g_2 \in G`, and such that:
+
+* **multiplication is associative**, :math:`g_1 \circ (g_2 \circ g_3) = (g_1 \circ g_2) \circ g_3` for all :math:`g_1, g_2, g_3 \in G`;
+* **there is a boring transformation** :math:`e` that does nothing, :math:`g \circ e = e \circ g = e` for all :math:`g \in G`;
+* **transformations can be undone**, with some :math:`g^{-1} \in G` such that :math:`g \circ g^{-1} = g^{-1} \circ g = e` for all :math:`g \in G`.
+
+It is sensible to ask that the unitary operators preserve the structure of the group:
+
+.. math::
+
+
+   U(g_1)U(g_2) = U(g_1 \circ g_2).
+
+For more on groups and how to represent them with matrices, see our `demo on geometric learning <https://pennylane.ai/qml/demos/tutorial_geometric_qml.html>`__.
+For the Hamiltonian to respect the symmetries encoded in the group :math:`G` means that it commutes with the matrices,
+
+.. math::
+
+
+   [U(g),\hat{H}] = 0
+
+for all :math:`g \in G`.
+Since the Hamiltonian generates time evolution, this means that if we
+apply a group transformation now or we apply it later, the effect is the
+same. Thus, we seek an algorithm which checks if these commutators are
+zero.
+
+Averaging over symmetries
+-----------------------------
+
+To verify that the Hamiltonian is symmetric with respect to :math:`G`, it
+seems like we will need to check each element
+:math:`g \in G` separately. There is a clever way to avoid this and boil
+it all down to a single number. For now, we’ll content ourselves with
+looking at the *average* over the group:
+
+.. math:: \frac{1}{|G|}\sum_{g\in G}[U(g),\hat{H}] = 0. 
+
+To make things concrete, let’s consider the `cyclic group <https://en.wikipedia.org/wiki/Cyclic_group>`__
+:math:`G = \mathbb{Z}_4`, which we can think of as rotations of the square. If we place a qubit
+on each corner, this group will naturally act on four qubits.
+
+.. figure:: ../demonstrations/testing_symmetry/square.png
+   :alt: square
+   :align: center
+   :width: 30%
+
+It is generated by a single rotation, which we’ll call :math:`c`. We’ll
+consider three Hamiltonians: one which is exactly symmetric with respect
+to :math:`G` (called :math:`\hat{H}_1`), one which is almost symmetric
+(called :math:`\hat{H}_2`), and one which isn’t at all symmetric
+(called :math:`\hat{H}_3`). We'll define them by
```

Comment:
```suggestion
consider three Hamiltonians: one which is exactly symmetric with respect
to :math:`G` (called :math:`\hat{H}_{\text{symm}}`), one which is almost symmetric
(called :math:`\hat{H}_{\tilde\text{symm}}`), and one which isn’t symmetric at all 
(called :math:`\hat{H}_{\text{asymm}}`). We'll define them by
```

Might be a bit more legible if the Hamiltonians had more descriptive subscripts. E.g., 

- H_1 --> H_{Symm}
- H_2 --> H_{~Symm}
- H_3 --> H_{ASymm}

Reply:
Good idea!